Fix failing unit tests on PHP 8.1

anton-vlasenko opened the pull request:
This PR aims to fix errors caused by incorrect usage of the `strncmp` function inside the `WP_Theme_JSON::get_property_value` methodon PHP 8.1 and above.
Please see the list of the errors below:

```
1) Tests_Blocks_Editor::test_get_block_editor_settings_theme_json_settings
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
2) Tests_Blocks_Editor::test_get_block_editor_settings_deprecated_filter_post_editor
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
12) Tests_Theme_wpGetGlobalStylesheet::test_block_theme_using_defaults
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
13) Tests_Theme_wpGetGlobalStylesheet::test_variables_in_classic_theme_with_no_presets_using_defaults
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
14) Tests_Theme_wpGetGlobalStylesheet::test_variables_in_classic_theme_with_presets_using_defaults
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
15) Tests_Theme_wpThemeJson::test_get_stylesheet_support_for_shorthand_and_longhand_values
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
16) Tests_Theme_wpThemeJson::test_get_stylesheet_skips_disabled_protected_properties
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
17) Tests_Theme_wpThemeJson::test_get_stylesheet_renders_enabled_protected_properties
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
18) Tests_Theme_wpThemeJson::test_get_stylesheet
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
19) Tests_Theme_wpThemeJson::test_get_stylesheet_preset_rules_come_after_block_rules
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
20) Tests_Theme_wpThemeJson::test_get_stylesheet_preset_values_are_marked_as_important
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
21) Tests_Theme_wpThemeJson::test_get_stylesheet_handles_whitelisted_element_pseudo_selectors
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
22) Tests_Theme_wpThemeJson::test_get_stylesheet_handles_only_pseudo_selector_rules_for_given_property
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
23) Tests_Theme_wpThemeJson::test_get_stylesheet_ignores_pseudo_selectors_on_non_whitelisted_elements
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
24) Tests_Theme_wpThemeJson::test_get_stylesheet_ignores_non_whitelisted_pseudo_selectors
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
25) Tests_Theme_wpThemeJson::test_get_stylesheet_handles_priority_of_elements_vs_block_elements_pseudo_selectors
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
26) Tests_Theme_wpThemeJson::test_get_stylesheet_handles_whitelisted_block_level_element_pseudo_selectors
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
27) Tests_Theme_wpThemeJson::test_remove_insecure_properties_removes_unsafe_styles
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
28) Tests_Theme_wpThemeJson::test_remove_insecure_properties_removes_unsafe_styles_sub_properties
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
29) Tests_Theme_wpThemeJson::test_remove_insecure_properties_applies_safe_styles
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
30) Tests_Theme_wpThemeJson::test_remove_invalid_element_pseudo_selectors
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
31) Tests_Theme_wpThemeJson::test_get_property_value_valid
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
32) Tests_Theme_wpThemeJson::test_get_property_value_loop
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
33) Tests_Theme_wpThemeJson::test_get_property_value_recursion
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
34) Tests_Theme_wpThemeJson::test_get_property_value_self
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
35) Tests_Theme_wpThemeJson::test_get_stylesheet_generates_layout_styles with data set "layout definitions" (array(array('default', 'flow', 'is-layout-flow', array(array(' > .alignleft', array('left', '0', '2em')), array(' > .alignright', array('right', '2em', '0')), array(' > .aligncenter', array('auto !important', 'auto !important'))), array(array(' > *', array('0', '0')), array(' > * + *', array(null, '0')))), array('flex', 'flex', 'is-layout-flex', 'flex', array(array('', array('wrap', 'center'))), array(array('', array(null))))))
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
36) Tests_Theme_wpThemeJson::test_get_stylesheet_generates_layout_styles_with_spacing_presets with data set "layout definitions" (array(array('default', 'flow', 'is-layout-flow', array(array(' > .alignleft', array('left', '0', '2em')), array(' > .alignright', array('right', '2em', '0')), array(' > .aligncenter', array('auto !important', 'auto !important'))), array(array(' > *', array('0', '0')), array(' > * + *', array(null, '0')))), array('flex', 'flex', 'is-layout-flex', 'flex', array(array('', array('wrap', 'center'))), array(array('', array(null))))))
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
37) Tests_Theme_wpThemeJson::test_get_stylesheet_generates_fallback_gap_layout_styles with data set "layout definitions" (array(array('default', 'flow', 'is-layout-flow', array(array(' > .alignleft', array('left', '0', '2em')), array(' > .alignright', array('right', '2em', '0')), array(' > .aligncenter', array('auto !important', 'auto !important'))), array(array(' > *', array('0', '0')), array(' > * + *', array(null, '0')))), array('flex', 'flex', 'is-layout-flex', 'flex', array(array('', array('wrap', 'center'))), array(array('', array(null))))))
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
38) Tests_Theme_wpThemeJson::test_get_stylesheet_generates_valid_block_gap_values_and_skips_null_or_false_values with data set "layout definitions" (array(array('default', 'flow', 'is-layout-flow', array(array(' > .alignleft', array('left', '0', '2em')), array(' > .alignright', array('right', '2em', '0')), array(' > .aligncenter', array('auto !important', 'auto !important'))), array(array(' > *', array('0', '0')), array(' > * + *', array(null, '0')))), array('flex', 'flex', 'is-layout-flex', 'flex', array(array('', array('wrap', 'center'))), array(array('', array(null))))))
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
39) Tests_Theme_wpThemeJson::test_get_styles_for_block_with_padding_aware_alignments
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
40) Tests_Theme_wpThemeJson::test_get_styles_for_block_without_padding_aware_alignments
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
41) Tests_Theme_wpThemeJson::test_get_styles_for_block_with_content_width
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
42) Tests_Webfonts_wpThemeJsonWebfontsHandler::test_font_face_generated_from_themejson
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
43) Tests_Webfonts_wpThemeJsonWebfontsHandler::test_font_face_not_generated with data set "no "fontFace" in theme.json" ('block-theme')
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
44) Tests_Webfonts_wpThemeJsonWebfontsHandler::test_font_face_not_generated with data set "empty "fontFace" in theme.json" ('empty-fontface-theme')
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
```

Trac ticket: https://core.trac.wordpress.org/ticket/56620



---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

jrfnl commented:
@anton-vlasenko Please open a separate PR for each individual fix using atomic commits, i.e. fix + associated tests in one commit.
This PR is trying to do too much, which makes it unclear what is being addressed by it and why.

jrfnl commented on tests/phpunit/tests/theme/wpThemeJson.php:
```diff
@@ -3874,4 +3874,19 @@ function data_set_spacing_sizes_when_invalid() {
+	/**
+	 * Tests that get_property_value static method returns an empty string
+	 * if the path is invalid.
+	 *
```
You may want to mention in this docblock that this test also safeguards that no "PHP 8.1 passing null to non-nullable deprecation" is thrown.

jrfnl replied:
Missing: the test should get a `@covers WP_Theme_JSON::get_property_value` tag.

anton-vlasenko replied:
Fixed in c58cc83386dbeba96f6527710331a00262f7607a, 6de32fdc203f91ab4ff713fb4b5903a424ae17be

anton-vlasenko committed:
Fix errors in unit tests:
1) Tests_Blocks_Editor::test_get_block_editor_settings_theme_json_settings
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
2) Tests_Blocks_Editor::test_get_block_editor_settings_deprecated_filter_post_editor
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
12) Tests_Theme_wpGetGlobalStylesheet::test_block_theme_using_defaults
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
13) Tests_Theme_wpGetGlobalStylesheet::test_variables_in_classic_theme_with_no_presets_using_defaults
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
14) Tests_Theme_wpGetGlobalStylesheet::test_variables_in_classic_theme_with_presets_using_defaults
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
15) Tests_Theme_wpThemeJson::test_get_stylesheet_support_for_shorthand_and_longhand_values
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
16) Tests_Theme_wpThemeJson::test_get_stylesheet_skips_disabled_protected_properties
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
17) Tests_Theme_wpThemeJson::test_get_stylesheet_renders_enabled_protected_properties
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
18) Tests_Theme_wpThemeJson::test_get_stylesheet
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
19) Tests_Theme_wpThemeJson::test_get_stylesheet_preset_rules_come_after_block_rules
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
20) Tests_Theme_wpThemeJson::test_get_stylesheet_preset_values_are_marked_as_important
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
21) Tests_Theme_wpThemeJson::test_get_stylesheet_handles_whitelisted_element_pseudo_selectors
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
22) Tests_Theme_wpThemeJson::test_get_stylesheet_handles_only_pseudo_selector_rules_for_given_property
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
23) Tests_Theme_wpThemeJson::test_get_stylesheet_ignores_pseudo_selectors_on_non_whitelisted_elements
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
24) Tests_Theme_wpThemeJson::test_get_stylesheet_ignores_non_whitelisted_pseudo_selectors
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
25) Tests_Theme_wpThemeJson::test_get_stylesheet_handles_priority_of_elements_vs_block_elements_pseudo_selectors
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
26) Tests_Theme_wpThemeJson::test_get_stylesheet_handles_whitelisted_block_level_element_pseudo_selectors
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
27) Tests_Theme_wpThemeJson::test_remove_insecure_properties_removes_unsafe_styles
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
28) Tests_Theme_wpThemeJson::test_remove_insecure_properties_removes_unsafe_styles_sub_properties
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
29) Tests_Theme_wpThemeJson::test_remove_insecure_properties_applies_safe_styles
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
30) Tests_Theme_wpThemeJson::test_remove_invalid_element_pseudo_selectors
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
31) Tests_Theme_wpThemeJson::test_get_property_value_valid
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
32) Tests_Theme_wpThemeJson::test_get_property_value_loop
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
33) Tests_Theme_wpThemeJson::test_get_property_value_recursion
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
34) Tests_Theme_wpThemeJson::test_get_property_value_self
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
35) Tests_Theme_wpThemeJson::test_get_stylesheet_generates_layout_styles with data set "layout definitions" (array(array('default', 'flow', 'is-layout-flow', array(array(' > .alignleft', array('left', '0', '2em')), array(' > .alignright', array('right', '2em', '0')), array(' > .aligncenter', array('auto !important', 'auto !important'))), array(array(' > *', array('0', '0')), array(' > * + *', array(null, '0')))), array('flex', 'flex', 'is-layout-flex', 'flex', array(array('', array('wrap', 'center'))), array(array('', array(null))))))
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
36) Tests_Theme_wpThemeJson::test_get_stylesheet_generates_layout_styles_with_spacing_presets with data set "layout definitions" (array(array('default', 'flow', 'is-layout-flow', array(array(' > .alignleft', array('left', '0', '2em')), array(' > .alignright', array('right', '2em', '0')), array(' > .aligncenter', array('auto !important', 'auto !important'))), array(array(' > *', array('0', '0')), array(' > * + *', array(null, '0')))), array('flex', 'flex', 'is-layout-flex', 'flex', array(array('', array('wrap', 'center'))), array(array('', array(null))))))
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
37) Tests_Theme_wpThemeJson::test_get_stylesheet_generates_fallback_gap_layout_styles with data set "layout definitions" (array(array('default', 'flow', 'is-layout-flow', array(array(' > .alignleft', array('left', '0', '2em')), array(' > .alignright', array('right', '2em', '0')), array(' > .aligncenter', array('auto !important', 'auto !important'))), array(array(' > *', array('0', '0')), array(' > * + *', array(null, '0')))), array('flex', 'flex', 'is-layout-flex', 'flex', array(array('', array('wrap', 'center'))), array(array('', array(null))))))
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
38) Tests_Theme_wpThemeJson::test_get_stylesheet_generates_valid_block_gap_values_and_skips_null_or_false_values with data set "layout definitions" (array(array('default', 'flow', 'is-layout-flow', array(array(' > .alignleft', array('left', '0', '2em')), array(' > .alignright', array('right', '2em', '0')), array(' > .aligncenter', array('auto !important', 'auto !important'))), array(array(' > *', array('0', '0')), array(' > * + *', array(null, '0')))), array('flex', 'flex', 'is-layout-flex', 'flex', array(array('', array('wrap', 'center'))), array(array('', array(null))))))
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
39) Tests_Theme_wpThemeJson::test_get_styles_for_block_with_padding_aware_alignments
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
40) Tests_Theme_wpThemeJson::test_get_styles_for_block_without_padding_aware_alignments
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
41) Tests_Theme_wpThemeJson::test_get_styles_for_block_with_content_width
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
42) Tests_Webfonts_wpThemeJsonWebfontsHandler::test_font_face_generated_from_themejson
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
43) Tests_Webfonts_wpThemeJsonWebfontsHandler::test_font_face_not_generated with data set "no "fontFace" in theme.json" ('block-theme')
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
44) Tests_Webfonts_wpThemeJsonWebfontsHandler::test_font_face_not_generated with data set "empty "fontFace" in theme.json" ('empty-fontface-theme')
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated

jrfnl commented on tests/phpunit/tests/theme/wpThemeJson.php:
```diff
@@ -3874,4 +3874,19 @@ function data_set_spacing_sizes_when_invalid() {
+		$reflection_class = new ReflectionClass( WP_Theme_JSON::class );
+
+		$get_property_value_method = $reflection_class->getMethod( 'get_property_value' );
+		$get_property_value_method->setAccessible( true );
+		$result = $get_property_value_method->invoke( null, array(), array( 'non_existent_path' ) );
```
Not strictly necessary, but good practice - reset the method to its original state after the invocation (and before the assertion):

```suggestion
		$get_property_value_method->setAccessible( true );
		$result = $get_property_value_method->invoke( null, array(), array( 'non_existent_path' ) );
		$get_property_value_method->setAccessible( false );
```

anton-vlasenko replied:
Fixed in af7d40a682a9e7d6bd0406febf536aa90054cabb

jrfnl reviewed (commented):
Thanks for setting up this patch & the associated test @anton-vlasenko !

All looks good and I have confirmed that the test would throw an error on PHP 8.1 without the patch and no longer does so with the patch.

I just left some small remarks about the test itself, mostly just nitpicks.

anton-vlasenko committed:
Improve the test.

jrfnl commented on src/wp-includes/class-wp-theme-json.php:
```diff
@@ -1717,7 +1717,12 @@ protected static function compute_style_properties( $styles, $settings = array()
+		$value = _wp_array_get( $styles, $path, '' );
+
+		// It doesn't make sense to process the value further.
+		if ( in_array( $value, array( '', null ), true ) ) {
```
Looks like this was changed after I reviewed.

Please change it to a direct comparison for better performance, `in_array()` is notoriously slow.

```suggestion
		if ( null === $value || '' === $value ) {
```

anton-vlasenko replied:
Yes, I changed it because I noticed 2 additional tests that had to be fixed.

anton-vlasenko replied:
Fixed in 0d66f734d39e7013a5d31a5916fd1847f88bcc04

jrfnl commented on tests/phpunit/tests/theme/wpThemeJson.php:
```diff
@@ -3874,4 +3874,42 @@ function data_set_spacing_sizes_when_invalid() {
 		);
 	}
+
+	/**
```
I very much welcome this data provider and the additional test.

As a convention, data providers are normally directly _after_ the test though, not _before_.

anton-vlasenko replied:
My bad. Thanks for pointing it out.
Fixed in 0d66f734d39e7013a5d31a5916fd1847f88bcc04

jrfnl replied:
No worries, you can't know everything. Thanks for changing the method order!

jrfnl reviewed (commented):
Couple more remarks after I noticed the implementation had received some minor changes since I started my original review.

anton-vlasenko committed:
Fix errors in unit tests:
1) Tests_Blocks_Editor::test_get_block_editor_settings_theme_json_settings
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
2) Tests_Blocks_Editor::test_get_block_editor_settings_deprecated_filter_post_editor
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
12) Tests_Theme_wpGetGlobalStylesheet::test_block_theme_using_defaults
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
13) Tests_Theme_wpGetGlobalStylesheet::test_variables_in_classic_theme_with_no_presets_using_defaults
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
14) Tests_Theme_wpGetGlobalStylesheet::test_variables_in_classic_theme_with_presets_using_defaults
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
15) Tests_Theme_wpThemeJson::test_get_stylesheet_support_for_shorthand_and_longhand_values
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
16) Tests_Theme_wpThemeJson::test_get_stylesheet_skips_disabled_protected_properties
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
17) Tests_Theme_wpThemeJson::test_get_stylesheet_renders_enabled_protected_properties
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
18) Tests_Theme_wpThemeJson::test_get_stylesheet
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
19) Tests_Theme_wpThemeJson::test_get_stylesheet_preset_rules_come_after_block_rules
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
20) Tests_Theme_wpThemeJson::test_get_stylesheet_preset_values_are_marked_as_important
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
21) Tests_Theme_wpThemeJson::test_get_stylesheet_handles_whitelisted_element_pseudo_selectors
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
22) Tests_Theme_wpThemeJson::test_get_stylesheet_handles_only_pseudo_selector_rules_for_given_property
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
23) Tests_Theme_wpThemeJson::test_get_stylesheet_ignores_pseudo_selectors_on_non_whitelisted_elements
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
24) Tests_Theme_wpThemeJson::test_get_stylesheet_ignores_non_whitelisted_pseudo_selectors
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
25) Tests_Theme_wpThemeJson::test_get_stylesheet_handles_priority_of_elements_vs_block_elements_pseudo_selectors
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
26) Tests_Theme_wpThemeJson::test_get_stylesheet_handles_whitelisted_block_level_element_pseudo_selectors
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
27) Tests_Theme_wpThemeJson::test_remove_insecure_properties_removes_unsafe_styles
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
28) Tests_Theme_wpThemeJson::test_remove_insecure_properties_removes_unsafe_styles_sub_properties
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
29) Tests_Theme_wpThemeJson::test_remove_insecure_properties_applies_safe_styles
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
30) Tests_Theme_wpThemeJson::test_remove_invalid_element_pseudo_selectors
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
31) Tests_Theme_wpThemeJson::test_get_property_value_valid
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
32) Tests_Theme_wpThemeJson::test_get_property_value_loop
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
33) Tests_Theme_wpThemeJson::test_get_property_value_recursion
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
34) Tests_Theme_wpThemeJson::test_get_property_value_self
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
35) Tests_Theme_wpThemeJson::test_get_stylesheet_generates_layout_styles with data set "layout definitions" (array(array('default', 'flow', 'is-layout-flow', array(array(' > .alignleft', array('left', '0', '2em')), array(' > .alignright', array('right', '2em', '0')), array(' > .aligncenter', array('auto !important', 'auto !important'))), array(array(' > *', array('0', '0')), array(' > * + *', array(null, '0')))), array('flex', 'flex', 'is-layout-flex', 'flex', array(array('', array('wrap', 'center'))), array(array('', array(null))))))
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
36) Tests_Theme_wpThemeJson::test_get_stylesheet_generates_layout_styles_with_spacing_presets with data set "layout definitions" (array(array('default', 'flow', 'is-layout-flow', array(array(' > .alignleft', array('left', '0', '2em')), array(' > .alignright', array('right', '2em', '0')), array(' > .aligncenter', array('auto !important', 'auto !important'))), array(array(' > *', array('0', '0')), array(' > * + *', array(null, '0')))), array('flex', 'flex', 'is-layout-flex', 'flex', array(array('', array('wrap', 'center'))), array(array('', array(null))))))
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
37) Tests_Theme_wpThemeJson::test_get_stylesheet_generates_fallback_gap_layout_styles with data set "layout definitions" (array(array('default', 'flow', 'is-layout-flow', array(array(' > .alignleft', array('left', '0', '2em')), array(' > .alignright', array('right', '2em', '0')), array(' > .aligncenter', array('auto !important', 'auto !important'))), array(array(' > *', array('0', '0')), array(' > * + *', array(null, '0')))), array('flex', 'flex', 'is-layout-flex', 'flex', array(array('', array('wrap', 'center'))), array(array('', array(null))))))
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
38) Tests_Theme_wpThemeJson::test_get_stylesheet_generates_valid_block_gap_values_and_skips_null_or_false_values with data set "layout definitions" (array(array('default', 'flow', 'is-layout-flow', array(array(' > .alignleft', array('left', '0', '2em')), array(' > .alignright', array('right', '2em', '0')), array(' > .aligncenter', array('auto !important', 'auto !important'))), array(array(' > *', array('0', '0')), array(' > * + *', array(null, '0')))), array('flex', 'flex', 'is-layout-flex', 'flex', array(array('', array('wrap', 'center'))), array(array('', array(null))))))
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
39) Tests_Theme_wpThemeJson::test_get_styles_for_block_with_padding_aware_alignments
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
40) Tests_Theme_wpThemeJson::test_get_styles_for_block_without_padding_aware_alignments
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
41) Tests_Theme_wpThemeJson::test_get_styles_for_block_with_content_width
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
42) Tests_Webfonts_wpThemeJsonWebfontsHandler::test_font_face_generated_from_themejson
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
43) Tests_Webfonts_wpThemeJsonWebfontsHandler::test_font_face_not_generated with data set "no "fontFace" in theme.json" ('block-theme')
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated
44) Tests_Webfonts_wpThemeJsonWebfontsHandler::test_font_face_not_generated with data set "empty "fontFace" in theme.json" ('empty-fontface-theme')
strncmp(): Passing null to parameter #1 ($string1) of type string is deprecated

anton-vlasenko committed:
Use direct comparison for performance reasons.

anton-vlasenko committed:
Improve the docblock.

anton-vlasenko committed:
Improve the test per this suggestion: https://github.com/WordPress/wordpress-develop/pull/3307#discussion_r983871718

anton-vlasenko commented:
> Thanks for setting up this patch & the associated test @anton-vlasenko !
> 
> All looks good and I have confirmed that the test would throw an error on PHP 8.1 without the patch and no longer does so with the patch.
> 
> I just left some small remarks about the test itself, mostly just nitpicks.

Thanks for the review, @jrfnl! I really appreciate it.
I've fixed the issues you mentioned.

jrfnl commented on src/wp-includes/class-wp-theme-json.php:
```diff
@@ -1717,7 +1717,12 @@ protected static function compute_style_properties( $styles, $settings = array()
+		$value = _wp_array_get( $styles, $path, '' );
+
+		// It doesn't make sense to process the value further.
+		if ( ( '' === $value ) || ( null === $value ) ) {
```
Nitpick: the extra parentheses are unnecessary here.

```suggestion
		if ( '' === $value || null === $value ) {
```

jrfnl reviewed (approved):
One last tiny nitpick but other than that, all good! ✅

anton-vlasenko commented:
> @anton-vlasenko Please open a separate PR for each individual fix using atomic commits, i.e. fix + associated tests in one commit.
> This PR is trying to do too much, which makes it unclear what is being addressed by it and why.

You are absolutely right. I've removed unrelated commits from this PR.
These unrelated commits were supposed to fix the `esc_url()` tests. I cannot reproduce these errors anymore (on the latest trunk).

jrfnl commented:
> > @anton-vlasenko Please open a separate PR for each individual fix using atomic commits, i.e. fix + associated tests in one commit.
> > This PR is trying to do too much, which makes it unclear what is being addressed by it and why.
> 
> You are absolutely right. I've removed unrelated commits from this PR. These unrelated commits were supposed to fix the `esc_url()` tests. I cannot reproduce these errors anymore (on the latest trunk).

I suspect that _might_ be related to two closely related patches for PHP 8.1 which got committed yesterday ([54349] and [54351]) (or some other PHP 8.1 patches committed earlier this week, so many moving parts... :joy:)

SergeyBiryukov commented:
Thanks for the PR! Merged in [r54362](https://core.trac.wordpress.org/changeset/54362).
